tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+    }
+
+    /// Obtains a reference to a service in the ready set by key.
+    pub fn get_ready<Q: Hash + Equivalent<K>>(&self, key: &Q) -> Option<(usize, &K, &S)> {
+        self.ready.get_full(key)
+    }
+
+    /// Obtains a mutable reference to a service in the ready set by key.
+    pub fn get_ready_mut<Q: Hash + Equivalent<K>>(
+        &mut self,
+        key: &Q,
+    ) -> Option<(usize, &K, &mut S)> {
+        self.ready.get_full_mut(key)
+    }
+
+    /// Obtains a reference to a service in the ready set by index.
+    pub fn get_ready_index(&self, idx: usize) -> Option<(&K, &S)> {
+        self.ready.get_index(idx)
+    }
+
+    /// Obtains a mutable reference to a service in the ready set by index.
+    pub fn get_ready_index_mut(&mut self, idx: usize) -> Option<(&mut K, &mut S)> {
+        self.ready.get_index_mut(idx)
+    }
+
+    /// Evicts an item from the cache.
+    ///
+    /// Returns true if a service was marked for eviction. The inner service may
+    /// be retained until `poll_pending` is called.
+    pub fn evict<Q: Hash + Equivalent<K>>(&mut self, key: &Q) -> bool {
+        let canceled = if let Some(c) = self.cancelations.swap_remove(key) {
+            c.send(()).expect("cancel receiver lost");
+            true
+        } else {
+            false
+        };
+
+        match self.ready.swap_remove_full(key) {
+            None => canceled,
+            Some((idx, _, _)) => {
+                self.repair_next_ready_index(idx);
+                true
+            }
+        }
+    }
+
+    // Updates `next_ready_index_idx` after the entry at `rm_idx` was evicted.
+    fn repair_next_ready_index(&mut self, rm_idx: usize) {
+        self.next_ready_index = self.next_ready_index.take().and_then(|idx| {
+            let new_sz = self.ready.len();
+            debug_assert!(idx <= new_sz && rm_idx <= new_sz);
+            let repaired = match idx {
+                i if i == rm_idx => None,         // removed
+                i if i == new_sz => Some(rm_idx), // swapped
+                i => Some(i),                     // uneffected
+            };
+            trace!(
+                "repair_index: orig={}; rm={}; sz={}; => {:?}",
+                idx,
+                rm_idx,
+                new_sz,
+                repaired,
+            );
+            repaired
+        });
+    }
+}
+
+impl<K, S, Req> ReadyCache<K, S, Req>
+where
+    K: Clone + Eq + Hash,
+    S: Service<Req>,
+    <S as Service<Req>>::Error: Into<error::Error>,
+    S::Error: Into<error::Error>,
+{
+    /// Pushes a new service onto the unready set.
+    ///
+    /// The service will be promoted to the ready set as `poll_pending` is invoked.
+    ///
+    /// Note that this does not replace a service currently in the ready set with
+    /// an equivalent key until this service becomes ready or the existing
+    /// service becomes unready.
+    pub fn push_service(&mut self, key: K, svc: S) {
+        let (tx, cancel) = oneshot::channel();
```

Comment:
I wonder if there's a way we can avoid creating a new oneshot each time a service is pushed. Might be tricky without exposing the inner wrapper though.

Reply:
Maybe by having a private function that pushes an `EvictableReady` directly that we use internally (in `call_next_ready` for example), and then a `pub` like this one that creates a new one.

Reply:
It probably would be possible if we don't use `FuturesUnordered` and instead implement something custom.

Reply:
I modified the ready-cache to only create a single oneshot per service, reusing it throughout its lifetime. This should generally amortize the cost of using a oneshot here.